### PR TITLE
Make metadata serchable

### PR DESCRIPTION
**Pull request by sergeysenja1992**

### CODE REVIEW

File: src/main/java/com/icthh/xm/ms/balance/config/jsonb/CustomPostgreSQL82Dialect.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+import java.sql.Types;
+import javax.persistence.criteria.CriteriaBuilder;
+import org.hibernate.dialect.function.SQLFunctionTemplate;
+import org.hibernate.type.StringType;
+
+public class CustomPostgreSQL82Dialect extends FixedPostgreSQL82Dialect {
+
+    public CustomPostgreSQL82Dialect() {
+        super();
+        registerFunction("json_field_text", new SQLFunctionTemplate(StringType.INSTANCE, "?1 ->> ?2"));
```
Comment (amedvedchuk) [inline src/main/java/com/icthh/xm/ms/balance/config/jsonb/CustomPostgreSQL82Dialect.java]: Extract to constants

File: src/main/java/com/icthh/xm/ms/balance/config/jsonb/JsonbUtils.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import javax.persistence.criteria.Expression;
+import javax.persistence.criteria.Path;
+import javax.persistence.criteria.Root;
+import lombok.experimental.UtilityClass;
+
+@UtilityClass
+public class JsonbUtils {
+
+    public static Expression<Integer> jsonIntField(CriteriaBuilder cb, Path<Map<String, String>> path, String fieldName) {
+        return cb.function("json_field_int", Integer.class, path, cb.literal(fieldName));
```
Comment (amedvedchuk) [inline src/main/java/com/icthh/xm/ms/balance/config/jsonb/JsonbUtils.java]: Use function name from constants

File: src/main/java/com/icthh/xm/ms/balance/domain/converter/MapToStringConverter.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import javax.persistence.AttributeConverter;
+import javax.persistence.Converter;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+
+@Slf4j
+@Converter
+public class MapToStringConverter implements AttributeConverter<Map<String, Object>, String> {
```
Comment (amedvedchuk) [inline src/main/java/com/icthh/xm/ms/balance/domain/converter/MapToStringConverter.java:18]: i see the same class in Entity.
maybe it worth to add it to some xm-commons sub-project?

File: src/main/resources/config/liquibase/changelog/20190815000000_make_matadata_jsonb.xml [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+            </column>
+        </addColumn>
+
+        <addColumn tableName="pocket_change_event">
+            <column name="metadata_json" type="${jsonType}">
+                <constraints nullable="true"/>
+            </column>
+        </addColumn>
+
+        <update tableName="pocket">
```
Comment (amedvedchuk) [inline src/main/resources/config/liquibase/changelog/20190815000000_make_matadata_jsonb.xml:36]: Suggest to separate DDL from DML in Liquibase script.

File: src/main/java/com/icthh/xm/ms/balance/domain/converter/MapToStringConverter.java (hunk 3 above)
Comment (sergeysenja1992) [reply src/main/java/com/icthh/xm/ms/balance/domain/converter/MapToStringConverter.java]: describe what you mean?

File: src/test/java/com/icthh/xm/ms/balance/repository/PostgressConfigIntTest.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+        Balance balance = balanceRepository.save(new Balance().typeKey("BALANCE").key("8").entityId(8L));
+        Pocket falsePocket = pocketRepository.save(pocket(18L, balance, of("bvalue", "false")));
+        Pocket truePocket = pocketRepository.save(pocket(28L, balance, of("bvalue", "true")));
+        Pocket first3Pocket = pocketRepository.save(pocket(38L, balance, of("eId", "3")));
+        Pocket second3Pocket = pocketRepository.save(pocket(48L, balance, of("eId", "3")));
+        pocketRepository.save(pocket(58L, balance, of("eId", "4")));
+        pocketRepository.save(pocket(68L, balance, of("eId", "4")));
+
+        List<Pocket> pockets3 = pocketRepository.findAll(Specification.where(
```
Comment (amedvedchuk) [inline src/test/java/com/icthh/xm/ms/balance/repository/PostgressConfigIntTest.java:109]: Is it possible to filter inside json not on the first level field?
for example if we have a pocket 
```
{
    "metadata_json": {
            "bvalue": true,
            "inner_container": {
                    "inValue": 123
             }
     }
}
```

how can we filter by `inValue`?

File: src/main/java/com/icthh/xm/ms/balance/domain/converter/MapToStringConverter.java (hunk 3 above)
Comment (sergeysenja1992) [reply src/main/java/com/icthh/xm/ms/balance/domain/converter/MapToStringConverter.java:18]: обсудили

**Review by dtsarenko — APPROVED** [review]

**Review by amedvedchuk — APPROVED** [review]

**Pull request merged**